File: dynamic-programming/grid/minimum-falling-path-sum/ts/solution_dfs.ts
const NOT_CALCULATED_VALUE = undefined

function minFallingPathSum(matrix: number[][]): number {
    const m = matrix.length, n = matrix[0].length
    const memory: number[][] = Array.from({length: m}, (): number[] => Array(n).fill(NOT_CALCULATED_VALUE))
    const dfs = (i: number, j: number): number => {
        if (j < 0 || j >= n) {
            return Number.MAX_SAFE_INTEGER
        }
        if (i === 0) {
            return matrix[0][j]
        }
        if (memory[i][j] !== NOT_CALCULATED_VALUE) {
            return memory[i][j]
        }
        return memory[i][j] = Math.min(dfs(i - 1, j - 1), dfs(i - 1, j), dfs(i - 1, j + 1)) + matrix[i][j]
    }
    let minSum: number = Number.MAX_SAFE_INTEGER
    for (let i = 0; i < n; i++) {
        minSum = Math.min(minSum, dfs(m - 1, i))
    }
    return minSum
}

export { minFallingPathSum }

File: dynamic-programming/grid/minimum-falling-path-sum/java/Solution_Dfs.java
import java.util.Arrays;
import java.util.function.BiFunction;

class Solution_Dfs {
    private static final Integer NOT_CALCULATED_VALUE = null;

    public int minFallingPathSum(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        Integer[][] memory = new Integer[m][n];
        for (Integer[] row : memory) {
            Arrays.fill(row, NOT_CALCULATED_VALUE);
        }
        BiFunction<Integer, Integer, Integer> dfs = new BiFunction<Integer, Integer, Integer>() {
            @Override
            public Integer apply(Integer i, Integer j) {
                if (j < 0 || j >= n) {
                    return Integer.MAX_VALUE;
                }
                if (i == 0) {
                    return matrix[0][j];
                }
                if (memory[i][j] != NOT_CALCULATED_VALUE) {
                    return memory[i][j];
                }
                return memory[i][j] = Math.min(Math.min(this.apply(i - 1, j - 1), this.apply(i - 1, j))
                        , this.apply(i - 1, j + 1)) + matrix[i][j];
            }
        };
        int minSum = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            minSum = Math.min(minSum, dfs.apply(m - 1, i));
        }
        return minSum;
    }
}